src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const stateNewObject = Object.assign({}, state);
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(stateNewObject, action.extraData);
+      stateHistory.push(Object.assign({}, stateNewObject));
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const item of action.keysToRemove) {
+        delete stateNewObject[item];
+      }
+      stateHistory.push(Object.assign({}, stateNewObject));
+    }
+
+    if (action.type === 'clear') {
+      for (const key of Object.keys(stateNewObject)) {
+        delete stateNewObject[key];
+      }
+      stateHistory.push(Object.assign({}, stateNewObject));
+    }
```

Comment:
Instead of using multiple `if` statements, consider using a `switch` statement to handle different action types. This will improve readability and maintainability.